kafka-client/src/main/java/dev/responsive/kafka/clients/ResponsiveRestoreConsumer.java

```diff
@@ -1,61 +1,114 @@
 import org.apache.kafka.clients.consumer.OffsetAndMetadata;
 import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.utils.LogContext;
+import org.slf4j.Logger;
 
 /**
  * Consumer implementation to use for restoring Responsive stores. This Restore Consumer emulates
- * consumping a topic that has been truncated up to some offset on a subset of topic partitions.
+ * consuming a topic that has been truncated up to some offset on a subset of topic partitions.
  * The specific offsets are provided by a function mapping the topic partition to the start
  * offset. This consumer is used in conjunction with Responsive Stores, which provide a safe
  * starting offset that has been committed to a remote storage system. This is useful for
  * minimizing the length of restores when using a Responsive Store over a compacted changelog
  * topic (which cannot be truncated).
+ * <p>
+ * This consumer is used by Kafka Streams for both active and standby task restoration, though
+ * only one or the other at a given time. This may be subject to change, but the general process
+ * follows this pattern:
+ * 1. initialize changelogs for newly assigned tasks:
+ *     i. restoreConsumer.assign(current assignment + new changelogs)
+ *     ii. if STANDBY_MODE: restoreConsumer.pause(standby changelogs)
+ *     iii. restoreConsumer.seekToBeginning(new changelogs)
+ *     iv. startOffset = restoreConsumer.position(changelog)
+ *     iv. for each ACTIVE changelog: call StateRestoreListener#onRestoreStart
+ * 2. restore active/standby state stores:
+ *     i. poll for new record batch
+ *     ii. restore batch through StateRestoreCallback
+ *     iii. for each ACTIVE changelog in batch: call StateRestoreListener#onBatchRestored
+ *     iv. update current offset based on highest offset in batch
  *
  * @param <K> Type of the record keys
  * @param <V> Type of the record values
  */
 public class ResponsiveRestoreConsumer<K, V> extends DelegatingConsumer<K, V> {
-  final Function<TopicPartition, OptionalLong> startOffsets;
+  private final Logger log;
+
+  private final Function<TopicPartition, OptionalLong> startOffsets;
+  private final Set<TopicPartition> uninitializedOffsets = new HashSet<>();
 
   public ResponsiveRestoreConsumer(
+      final String clientId,
       final Consumer<K, V> delegate,
       final Function<TopicPartition, OptionalLong> startOffsets
   ) {
     super(delegate);
     this.startOffsets = Objects.requireNonNull(startOffsets);
+    this.log = new LogContext(
+        String.format("responsive-restore-consumer [%s]", Objects.requireNonNull(clientId))
+    ).logger(ResponsiveConsumer.class);
   }
 
-  private List<TopicPartition> seekInitial(final Collection<TopicPartition> partitions) {
-    final List<TopicPartition> notFound = new LinkedList<>();
+  private Set<TopicPartition> initializeOffsets(final Collection<TopicPartition> partitions) {
+    final Set<TopicPartition> notFound = new LinkedHashSet<>();
```

Comment:
Ah, yeah this can just be a regular HashSet. I was trying to keep it List-like because that was the original type, but looking at how it's used, HashSet will be fine too